### PR TITLE
feat: atomic object update

### PR DESCRIPTION
**Pull request by alexgao001**
### Description

Enable object update. A user can update the  object in 2 steps:
1. Send **MsgUpdateObejctContent** to Greenfield
2. Put Object to SP.

Change: 

Greenfield chain introduces a new model **ShadowObjectInfo** for store user updated object meta temporarily. This info will serve until the updated object is sealed again, then it will be removed from chain.

and extra fileds are added to **ObjectInfo**
```
message ObjectInfo {
  ..... 
  // is_updating indicates whether a object is being updated.
  bool is_updating = 16;
  // updated_at define the block timestamp when the object is updated
  int64 updated_at = 17;
  // version define the version of object
  int64 version = 18;
}
```

SP will utilize these info to differentiate a user's request is creating a new object or updating a existing one. 

PieceKey: 
the **verison** is added to **ObjectInfo**, in order not to conflict with exiting piece. Any newer version of piece will append the version suffix. 

DB: **ShadowIntegrityMeta**
Used for the updated object meta. It will be copied to **IntegrityMeta** once the new object is sealed. 

GC: 
new GC stale version task is added. Since piece hash table will also be used for the updated object replication. Will skip the current GC of zombie piece if a object is updating status. 
  
More details please refer to https://github.com/bnb-chain/BEPs/pull/346

### Rationale

tell us why we need these changes...

### Example

add an example CLI or API response...

### Changes

Notable changes: 
* add each change in a bullet point here
* ...

### Potential Impacts
* add potential impacts for other components here
* ...

### CODE REVIEW

File: store/sqldb/shadow_object_integrity.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+	insertIntegrityMetaRecord := &ShadowIntegrityMetaTable{
+		ObjectID:          meta.ObjectID,
+		RedundancyIndex:   meta.RedundancyIndex,
+		PieceChecksumList: util.BytesSliceToString(meta.PieceChecksumList),
+		IntegrityChecksum: hex.EncodeToString(meta.IntegrityChecksum),
+		Version:           meta.Version,
+	}
+	result := s.db.Table(ShadowIntegrityMetaTableName).Create(insertIntegrityMetaRecord)
+	if result.Error != nil && MysqlErrCode(result.Error) == ErrDuplicateEntryCode {
+		return nil
```
Comment (clydemeng) [inline store/sqldb/shadow_object_integrity.go:59]: shall we log "duplicated setting request " and return err here?

File: modular/manager/manage_task.go [hunk 2]
```diff
@@ -10,6 +10,7 @@ import (
 	"time"
 
 	"cosmossdk.io/math"
+	"github.com/bnb-chain/greenfield-storage-provider/core/piecestore"
```
Comment (constwz) [inline modular/manager/manage_task.go:13]: goimports

File: store/sqldb/shadow_object_integrity.go (hunk 1 above)
Comment (clydemeng) [reply store/sqldb/shadow_object_integrity.go:61]: it might be useful for investigation if there is any issue

**Comment by clydemeng** [comment]
LGTM generally
please add UT

**Comment by ruojunm** [comment]
LGTM generally
please add UT

**Review by ruojunm — APPROVED** [review]

**Review by constwz — APPROVED** [review]
LGTM

File: modular/blocksyncer/modules/object/object_handler.go [hunk 3]
```diff
@@ -110,8 +116,28 @@ func (m *Module) ExtractEventStatements(ctx context.Context, block *tmctypes.Res
 			return nil, errors.New("update object event assert error")
 		}
 		return m.handleUpdateObjectInfo(ctx, block, txHash, updateObjectInfo), nil
+	case EventUpdateObjectContent:
```
[the remark this answers is not in the document]
Comment (alexgao001) [reply modular/blocksyncer/modules/object/object_handler.go:119]: added

File: modular/metadata/metadata_sp_exit_service.go [hunk 4]
```diff
@@ -162,7 +162,7 @@ func (r *MetadataModular) GfSpListMigrateBucketEvents(ctx context.Context, req *
 		return nil, err
 	}
 	if uint64(latestBlock) < req.BlockId {
-		log.CtxError(ctx, "failed to list migrate bucket events due to request block id exceed current block syncer block height")
+		//log.CtxError(ctx, "failed to list migrate bucket events due to request block id exceed current block syncer block height")
```
[the remark this answers is not in the document]
Comment (alexgao001) [reply modular/metadata/metadata_sp_exit_service.go]: reverted

**Pull request merged**